flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java

```diff
@@ -73,46 +79,150 @@ public static Object[] rowToArray(ResultSet rs, int size) throws SQLException {
+                            if (!hasPrimaryKey) {
+                                upsertBinlog(
+                                        snapshotRecords,
+                                        binlogRecord,
+                                        createReadOpValue(binlogRecord, Envelope.FieldName.BEFORE),
+                                        true);
+                                if (!splitKeyRangeContains(
+                                        getSplitKey(
+                                                splitBoundaryType, nameAdjuster, structFromAfter),
+                                        splitStart,
+                                        splitEnd)) {
+                                    LOG.warn(
+                                            "The updated chunk key is out of the split range. Cannot provide exactly-once semantics.");
+                                }
+                            }
+                            // If the chunk key changed, we still send here
+                            // This will cause the at-least-once semantics
+                            upsertBinlog(
+                                    snapshotRecords,
+                                    binlogRecord,
+                                    hasPrimaryKey ? keyStruct : structFromAfter,
+                                    false);
+                            break;
+                        case DELETE:
+                            upsertBinlog(
+                                    snapshotRecords,
+                                    binlogRecord,
+                                    hasPrimaryKey
+                                            ? keyStruct
+                                            : createReadOpValue(
+                                                    binlogRecord, Envelope.FieldName.BEFORE),
+                                    true);
+                            break;
+                        case READ:
+                            throw new IllegalStateException(
+                                    String.format(
+                                            "Binlog record shouldn't use READ operation, the the record is %s.",
+                                            binlogRecord));
+                    }
+                }
+            }
+        }
+    }
+
+    private static void upsertBinlog(
+            Map<Struct, List<SourceRecord>> snapshotRecords,
+            SourceRecord binlogRecord,
+            Struct keyStruct,
+            boolean isDelete) {
+        boolean hasPrimaryKey = binlogRecord.key() != null;
+        List<SourceRecord> records = snapshotRecords.get(keyStruct);
+        if (isDelete) {
+            if (records == null || records.isEmpty()) {
+                LOG.error(
+                        "Deleting a record which is not in its split for tables without primary keys. This may happen when the chunk key column is updated in another snapshot split.");
```

Comment:
we need open a doc PR to guide user how to choose chunk key column

Reply:
I will provide the docs about these experimental features later. See #2164.